### PR TITLE
fix(es/codegen): Skip `lo` of a program if body is not empty

### PR DESCRIPTION
**Pull request by kdy1**
**Description:**

**Related issue:**

 - Closes https://github.com/swc-project/swc/issues/6767.

### CODE REVIEW

**Review by kdy1 — COMMENTED** [review]
reset ci


---

swc-bump:
 - swc_ecma_codegen

**Comment by kdy1** [comment]
@jridgewell Do you think this is correct?
The actual problem of #6767 is that the `span.lo` of `Module` does not match `span.lo` of the first item.

**Review by jridgewell — DISMISSED** [review]

**Review by jridgewell — COMMENTED** [review]
I think this solves this particular case, but the larger solution would be to to mark `DUMMY_SPAN` locations with a source-less mapping. We'd need to remove the dummy checks from https://github.com/swc-project/swc/blob/b102504eb9e0afc5a79989c9cd8a3952ab1e2689/crates/swc_html_codegen/src/writer/basic.rs.

**Comment by kdy1** [comment]
I tried it, but it resulted in extremely huge source maps.
So I reverted it.

The problem was that so many nodes have `DUMMY_SP`

**Comment by jridgewell** [comment]
We can optimize source maps output, but we still need all locations:
- https://github.com/jridgewell/gen-mapping/blob/9ff6f25fa9bd12eeb9d265a17dd7631133669fdc/src/gen-mapping.ts#L420-L458
- https://github.com/jridgewell/babel/blob/518e1a6de4e05613ebf3aaf2c35e598e1dbcdb63/packages/babel-generator/src/source-map.ts#L82-L101

File: crates/swc_ecma_codegen/src/lib.rs [hunk 1]
```diff
@@ -94,7 +94,10 @@ where
     #[emitter]
     pub fn emit_module(&mut self, node: &Module) -> Result {
         self.emit_leading_comments_of_span(node.span(), false)?;
-        srcmap!(node, true);
+
+        if node.body.is_empty() {
```
Comment (kdy1) [inline crates/swc_ecma_codegen/src/lib.rs:98]: Note: This is still required to fix #6767

File: crates/swc/tests/fixture/sourcemap/issue-3854/1-true/output/index.map [hunk 2]
```diff
@@ -1,5 +1,5 @@
...
...qBAAA,CAiQe,IAAI,EAAEe,AAjQrB,YAAA,EAiQgCf,AAjQhC,KAAA,AAiQqC;IACjC;IAZAiB,YAAY,GAAG5B,AAtPnB,KAAA,AAsPkD,CAAE;QAC5C,KAAK;QAHT,AApPJ,sBAAA,CAAA,IAAA,EAoPI,AApPJ,YAAA;YAAA,QAAA,EAAA,IAAA;YAAA,KAAA,EAoP2C,IAAII,AApP/C,uBAAA;;QAwPQ,IAAI,CAACJ,AAxPb,KAAA,AAwPkB,IAAIA,AAxPtB,KAAA;IAyPI;AAUJ;IAGI,AAtQJ,OAAA,qBAAA,OAAA,IAuQI,AAvQJ,QAAA,qBAAA,OAAA;AAqQA,MAAME,AArQN,qBAAA,SAqQoCG,AArQpC,iBAAA,CAAA,eAAA,AAqQmD;IAI/CL,AAzQJ,KAAA,CAyQU,GAAGA,AAzQb,KAAA,AAyQ4C,EAAQ;QAzQpD,qBAAA,CAAA,IAAA,EA0QcA,AA1Qd,OAAA,EA0QsBA,AA1QtB,KAAA,CA0Q4BO,AA1Q5B,GAAA,AA0Q+B,CAACC,AA1QhC,aAAA,CAAA,WAAA,AA0Q2C;QACnC,OAAO,IAAI;IACf;IAEAK,AA9QJ,MAAA,CA8QW,GAAGA,AA9Qd,MAAA,AA8Q8C,EAAQ;QA9QtD,qBAAA,CAAA,IAAA,EA+QcA,AA/Qd,QAAA,EA+QuBA,AA/QvB,MAAA,CA+Q8BN,AA/Q9B,GAAA,AA+QiC,CAACC,AA/QlC,aAAA,CAAA,WAAA,AA+Q6C;QACrC,OAAO,IAAI;IACf;IAEAG,AAnRJ,KAAA,GAmR4B;QACpB,OAAO;YACHX,AArRZ,KAAA,AAqRiB,EArRjB,qBAAA,CAqRmB,IAAI,EAAEA,AArRzB,OAAA;YAsRYa,AAtRZ,MAAA,AAsRkB,EAtRlB,qBAAA,CAsRoB,IAAI,EAAEA,AAtR1B,QAAA;QAuRQ;IACJ;gBAxRJ,GAAA,IAAA;QAAA,KAAA,IAAA,IAAA;QAsQI,AAtQJ,sBAAA,CAAA,IAAA,EAsQI,AAtQJ,OAAA;YAAA,QAAA,EAAA,IAAA;YAAA,KAAA,EAsQ8B,EAAE;;QAC5B,AAvQJ,sBAAA,CAAA,IAAA,EAuQI,AAvQJ,QAAA;YAAA,QAAA,EAAA,IAAA;YAAA,KAAA,EAuQ+B,EAAE;;;AAkBjC",
```
Comment (kdy1) [inline crates/swc/tests/fixture/sourcemap/issue-3854/1-true/output/index.map:2]: @jridgewell Is this better than the previous one?

**Comment by kdy1** [comment]
I'll use a separate PR.

**Review by jridgewell — APPROVED** [review]

**Pull request merged**